src/rust/engine/process_execution/src/immutable_inputs.rs

```diff
@@ -0,0 +1,84 @@
+use std::collections::{BTreeMap, HashMap};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use double_checked_cell_async::DoubleCheckedCell;
+use fs::RelativePath;
+use hashing::Digest;
+use parking_lot::Mutex;
+use store::Store;
+use tempfile::TempDir;
+
+use crate::WorkdirSymlink;
+
+/// Holds Digests materialized into a temporary directory, for symlinking into local sandboxes.
+pub struct ImmutableInputs {
+  store: Store,
+  // The TempDir that digests are materialized in.
+  workdir: TempDir,
+  // A map from Digest to the location it has been materialized at. The DoubleCheckedCell allows
+  // for cooperation between threads attempting to create Digests.
+  contents: Mutex<HashMap<Digest, Arc<DoubleCheckedCell<PathBuf>>>>,
+}
+
+impl ImmutableInputs {
+  pub fn new(store: Store, base: &Path) -> Result<Self, String> {
+    let workdir = TempDir::new_in(base).map_err(|e| {
+      format!(
+        "Failed to create temporary directory for immutable inputs: {}",
+        e
+      )
+    })?;
+    Ok(Self {
+      store,
+      workdir,
+      contents: Mutex::default(),
+    })
+  }
+
+  /// Returns an absolute Path to immutably consume the given Digest from.
+  async fn path(&self, digest: Digest) -> Result<PathBuf, String> {
+    let cell = self.contents.lock().entry(digest).or_default().clone();
+    let value: Result<_, String> = cell
+      .get_or_try_init(async {
+        let digest_str = digest.hash.to_hex();
+        let digest_rel_dir = digest_str.chars().take(2).collect::<String>();
+
+        let path = self.workdir.path().join(digest_rel_dir).join(digest_str);
+        self
+          .store
+          .materialize_directory(path.clone(), digest)
+          .await?;
+        Ok(path)
```

Comment:
One potentially blocking issue is that we do not currently materialize the directory as read only. It might be worthwhile to do that before landing this.

Reply:
How would that be accomplished? Remove write bits on files materialized?

One thought I had (which would only work on Linux) is to bind mount the directory read-only instead of symlinking. But that is non-trivial to implement.

Reply:
I <3 macOS

Reply:
This change is now atop #13857.